File: three-number-sum/solution-1.js
array = [12, 3, 1, 2, -6, 5, -8, 6]
targetSum = 0;

function threeNumberSum(array, targetSum) {
    array.sort((a, b) => a - b); // sort the array desc to asc
    let triplets = [];
    // loop thru the array and stop when i is the 3rd to last index in the array
    for (let i = 0; i < array.length - 2; i++) {
        // i = current value of index
        let left = i + 1; // 1 position left of i (left value)
        let right = array.length - 1; // last value
        while (left < right) {
            const currentSum = array[i] + array[left] + array[right];
            if (currentSum === targetSum) {
                triplets.push([array[i], array[left++], array[right--]]);
            }
            if (currentSum < targetSum) left += 1;
            if (currentSum > targetSum) right -= 1;
        }
    }
    return triplets;
}

File: medium/three-number-sum/solution-1.py
# O(n^2) time && O(n) space

array = [12, 3, 1, 2, -6, 5, -8, 6]
targetSum = 0

def threeNumberSum(array, targetSum):
    array.sort()
    triples = []
    for i in range(len(array) - 2):
        left = i + 1
        right = len(array) - 1
        while left < right:
            currentSum = array[i] + array[left] + array[right]
            if currentSum == targetSum:
                triples.append([array[i], array[left], array[right]])
                left += 1
                right -= 1
            elif currentSum < targetSum:
                left += 1
            elif currentSum > targetSum:
                right -= 1
    return triples